LESQ-977: Redirect user through onboarding on signup

carlmw opened the pull request:
## Description

Music year: [1999](https://www.youtube.com/watch?v=9Ht5RZpzPqw)

- Redirects users to the onboarding page immediately after sign-up (or login if they have not previously completed onboarding)
- Redirects signed-in users to the onboarding page from any other page if they have not previously completed onboarding

## How to test

1. You'll need the use-auth-owa [feature flag enabled](https://eu.posthog.com/project/124/feature_flags/27071)
2. Go to https://deploy-preview-2681--oak-web-application.netlify.thenational.academy/teachers/programmes/art-secondary-ks3-l/units/pop-art-modern-popular-culture-4928/lessons/chila-kumari-burman-part-1-cdh3jc
3. Click on "Download resources"
4. You should see a login screen
5. Sign-up or login
6. You should see the onboarding page
7. Complete the onboarding flow
8. Click continue
3. You should see the download page and be able to download files

## Screenshots


https://github.com/user-attachments/assets/0caedc6b-b6c5-4469-8230-3ce81f116920



**Basic error state if the request to onboard the user fails (server or network error)**
I think this treatment can be improved but for now at least it gives some feedback instead of failing silently.
<img width="604" alt="Screenshot 2024-08-09 at 10 09 47" src="https://github.com/user-attachments/assets/a6401dbe-d27c-4fb7-974c-014e75cdd46f">

### For testing
You can trigger the onboarding flow for an existing user by:

1. Signing out if you're signed 
2. Finding the user in Clerk under the development app and removing `owa:onboarded` from public metadata

<img width="1199" alt="Screenshot 2024-08-19 at 09 44 46" src="https://github.com/user-attachments/assets/4b4de5e3-defe-4c9d-9c42-c4802e2d8977">


  

## Checklist

- [ ] Added or updated tests where appropriate
- [ ] Manually tested across browsers / devices
- [ ] Considered impact on accessibility
- [ ] Design sign-off
- [ ] Approved by product owner
- [ ] Does this PR update a package with a breaking change

carlmw commented on src/__tests__/__helpers__/renderWithProviders.tsx:
```diff
@@ -77,7 +77,7 @@ const providersByName: {
   toast: [ToastProvider],
   menu: [MenuProvider],
-  user: [UserProvider, { user: undefined }],
+  user: [UserProvider, { user: undefined, fetcher: async () => undefined }],
```
I realised that some tests were making a request to load the user, this stops that.

carlmw commented on src/node-lib/auth/userManagement.ts:
```diff
@@ -0,0 +1,10 @@
+
+import getServerConfig from "../getServerConfig";
+
+export const userManagementClient = new ManagementClient({
+  domain: getServerConfig("auth0IssuerBaseURL").replace("https://", ""),
+  clientId: getServerConfig("auth0UserManagementClientId"),
+  clientSecret: getServerConfig("auth0UserManagementClientSecret"),
+});
```
This is the "OWA user management" application in Auth0. It only has access to update user's app metadata. 

Keeping this isolated will make it easier to extract the whole thing into a separate identity service at a later date.

carlmw commented on src/pages/onboarding.tsx:
```diff
@@ -29,4 +30,4 @@ const OnboardingPage: NextPage = () => {
 };
 
-export default OnboardingPage;
+export default withPageAuthRequired(OnboardingPage);
```
The onboarding page now requires auth to function.

carlmw commented on src/app/api/auth/onboarding/route.ts:
```diff
@@ -0,0 +1,28 @@
+    throw new OakError({ code: "auth/missing-session", meta: { req } });
+  }
+
+  const appMetadata = { "owa:onboarded": true };
```
We can add extra data here from the onboarding form to mark the user as a teacher. For now it simply flips the user to the onboarded state.

carlmw commented on src/components/TeacherViews/Onboarding/Onboarding.view.tsx:
```diff
@@ -69,8 +75,23 @@ export const OnboardingView = () => {
+      // TODO: something with this data
+      console.log("onboarding form values: ", data);
+
+      await onboardUser();
```
We can post extra data here to mark the user as a teacher. For now this action simply flips the user to the onboarded state in Auth0.

carlmw commented on src/app/api/auth/[auth0]/route.ts:
```diff
@@ -12,4 +12,9 @@ export const GET = auth0.handleAuth({
   "oidc-logout": auth0.handleLogout({
     returnTo: oidcLogoutUrl.toString(),
   }),
+  "silent-login": auth0.handleLogin({
```
Sends the user back through login without displaying any UI. This is used to update the user's id and access tokens with any new claims following onboarding. Right now all this adds is the `owa:onboarded: true` claim to their id token which will show up in their user object. If they're not already logged in this will error.

carlmw commented on src/components/TeacherViews/Onboarding/Onboarding.test.tsx:
```diff
@@ -1,11 +1,36 @@
 describe("Onboarding view", () => {
+  beforeAll(() => {
+    fetchMock.enableMocks();
+    fetchMock.doMock(
+      JSON.stringify([
+        {
+          urn: "100224",
+          la: "Hackney",
+          name: "De Beauvoir Primary School",
+          postcode: "E8 3DY",
+          fullInfo: "100224, Hackney, De Beauvoir Primary School, E8 3DY",
+          status: "Open, but proposed to close",
+        },
+      ]),
+    );
+  });
```
The tests were hitting the real school picker API.

njblore commented on src/hooks/useRequireOnboarding.ts:
```diff
@@ -0,0 +1,29 @@
+ * Redirects the user to onboarding after login
+ * when they have not previously completed onboarding
+ */
+export function useRequireOnboarding() {
```
is this something we could do in middleware if we can figure out the env situation?

carlmw replied:
Yeah, that is what i'm hoping 👍🏻

mikeritson-oak commented:
### Testing Results
I've tested this on the preview branch in the initial comment of this PR. Three possible issues seen:

1. Visible flash (mentioned in this PR).
2. Onboarding loop (seen in other testing).
3. Missing icon on error state.

## Visible flash (mentioned in this PR)
As mentioned in the initial comment _"There is a visible jump when the user is redirected to the onboarding screen immediately after signing in."_, this is still happening at this point of development.

## Onboarding loop (seen in other testing)
If the user logins in and gets to the /onboarding page, but then subsequently clicks away, they are returned to /onboarding moments later and don't seem to have any way to navigate away.

## Missing icon on error state.
I think this one was only due to the method of testing, having no network connection to simulate a disconnect. As such, the icon couldn't load but on checking the SVG was present on the relevant URL.

![SCR-20240812-kyrj](https://github.com/user-attachments/assets/c8f9d173-11c6-4ae2-9dcf-9736b1e9e5af)

jack-skerrett-oak commented:
> ### Testing Results
> I've tested this on the preview branch in the initial comment of this PR. Three possible issues seen:
> 
> 1. Visible flash (mentioned in this PR).
> 2. Onboarding loop (seen in other testing).
> 3. Missing icon on error state.
> 
> ## Visible flash (mentioned in this PR)
> As mentioned in the initial comment _"There is a visible jump when the user is redirected to the onboarding screen immediately after signing in."_, this is still happening at this point of development.
> 
> ## Onboarding loop (seen in other testing)
> If the user logins in and gets to the /onboarding page, but then subsequently clicks away, they are returned to /onboarding moments later and don't seem to have any way to navigate away.
> 
> ## Missing icon on error state.
> I think this one was only due to the method of testing, having no network connection to simulate a disconnect. As such, the icon couldn't load but on checking the SVG was present on the relevant URL.
> 
> ![SCR-20240812-kyrj](https://private-user-images.githubusercontent.com/167184734/357086523-c8f9d173-11c6-4ae2-9dcf-9736b1e9e5af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NzM0MDksIm5iZiI6MTcyMzQ3MzEwOSwicGF0aCI6Ii8xNjcxODQ3MzQvMzU3MDg2NTIzLWM4ZjlkMTczLTExYzYtNGFlMi05ZGNmLTk3MzZiMWU5ZTVhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxMlQxNDMxNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MjdjMmU3MzRkYTFjZGJmZTE2MjczMThiODJiYTA2MTUyMTliZjVjNjgzM2M3YTVkZmM2NzU4YjFlNGZhZjE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-fwOQJxKlvyg0GsNK1YSlZuWLP3cN3k2DtCVSEkSg6c)

@carlmw that onboarding loop is particularly problematic because it locks the user from ever using the site (without doing some things that I doubt our typical user would want to do). I am assuming it is part of creating the auth token so it needs to be done before they can carry on their journey?

carlmw commented:
> > ### Testing Results
> > I've tested this on the preview branch in the initial comment of this PR. Three possible issues seen:
> > 
> > 1. Visible flash (mentioned in this PR).
> > 2. Onboarding loop (seen in other testing).
> > 3. Missing icon on error state.
> > 
> > ## Visible flash (mentioned in this PR)
> > As mentioned in the initial comment _"There is a visible jump when the user is redirected to the onboarding screen immediately after signing in."_, this is still happening at this point of development.
> > ## Onboarding loop (seen in other testing)
> > If the user logins in and gets to the /onboarding page, but then subsequently clicks away, they are returned to /onboarding moments later and don't seem to have any way to navigate away.
> > ## Missing icon on error state.
> > I think this one was only due to the method of testing, having no network connection to simulate a disconnect. As such, the icon couldn't load but on checking the SVG was present on the relevant URL.
> > ![SCR-20240812-kyrj](https://private-user-images.githubusercontent.com/167184734/357086523-c8f9d173-11c6-4ae2-9dcf-9736b1e9e5af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NzM0MDksIm5iZiI6MTcyMzQ3MzEwOSwicGF0aCI6Ii8xNjcxODQ3MzQvMzU3MDg2NTIzLWM4ZjlkMTczLTExYzYtNGFlMi05ZGNmLTk3MzZiMWU5ZTVhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxMlQxNDMxNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MjdjMmU3MzRkYTFjZGJmZTE2MjczMThiODJiYTA2MTUyMTliZjVjNjgzM2M3YTVkZmM2NzU4YjFlNGZhZjE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-fwOQJxKlvyg0GsNK1YSlZuWLP3cN3k2DtCVSEkSg6c)
> 
> @carlmw that onboarding loop is particularly problematic because it locks the user from ever using the site (without doing some things that I doubt our typical user would want to do) if they decide they don't want to continue the sign up journey. I am assuming it is part of creating the auth token so it needs to be done before they can carry on their journey?

@jack-skerrett-oak @mikeritson-oak at that point they are already signed-up and logged in. The intention there is to ensure that they complete the onboarding process before continuing to use the rest of the site. It's what i'm alluding to [here in Slack](https://oaknationalacademy.slack.com/archives/C03HZ76AE3Z/p1723197888223549). If you think that isn't correct we should probably raise it over on that thread as it's a product decision (I know there is a desire to capture this info). We could tweak it to only push the user through onboarding again when they try to access another page that requires authentication if that makes sense?

jack-skerrett-oak commented:
> > > ### Testing Results
> > > I've tested this on the preview branch in the initial comment of this PR. Three possible issues seen:
> > > 
> > > 1. Visible flash (mentioned in this PR).
> > > 2. Onboarding loop (seen in other testing).
> > > 3. Missing icon on error state.
> > > 
> > > ## Visible flash (mentioned in this PR)
> > > As mentioned in the initial comment _"There is a visible jump when the user is redirected to the onboarding screen immediately after signing in."_, this is still happening at this point of development.
> > > ## Onboarding loop (seen in other testing)
> > > If the user logins in and gets to the /onboarding page, but then subsequently clicks away, they are returned to /onboarding moments later and don't seem to have any way to navigate away.
> > > ## Missing icon on error state.
> > > I think this one was only due to the method of testing, having no network connection to simulate a disconnect. As such, the icon couldn't load but on checking the SVG was present on the relevant URL.
> > > ![SCR-20240812-kyrj](https://private-user-images.githubusercontent.com/167184734/357086523-c8f9d173-11c6-4ae2-9dcf-9736b1e9e5af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NzM0MDksIm5iZiI6MTcyMzQ3MzEwOSwicGF0aCI6Ii8xNjcxODQ3MzQvMzU3MDg2NTIzLWM4ZjlkMTczLTExYzYtNGFlMi05ZGNmLTk3MzZiMWU5ZTVhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxMlQxNDMxNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MjdjMmU3MzRkYTFjZGJmZTE2MjczMThiODJiYTA2MTUyMTliZjVjNjgzM2M3YTVkZmM2NzU4YjFlNGZhZjE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-fwOQJxKlvyg0GsNK1YSlZuWLP3cN3k2DtCVSEkSg6c)
> > 
> > 
> > @carlmw that onboarding loop is particularly problematic because it locks the user from ever using the site (without doing some things that I doubt our typical user would want to do) if they decide they don't want to continue the sign up journey. I am assuming it is part of creating the auth token so it needs to be done before they can carry on their journey?
> 
> @jack-skerrett-oak @mikeritson-oak at that point they are already signed-up and logged in. The intention there is to ensure that they complete the onboarding process before continuing to use the rest of the site. It's what i'm alluding to [here in Slack](https://oaknationalacademy.slack.com/archives/C03HZ76AE3Z/p1723197888223549). If you think that isn't correct we should probably raise it over on that thread as it's a product decision (I know there is a desire to capture this info). We could tweak it to only push the user through onboarding again when they try to access another page that requires authentication if that makes sense?

@carlmw that is how I interpreted that thread. I don't think we want to block users from accessing areas of the site that don't require signup because they haven't finished onboarding, but everywhere that does require signup to access (currently downloads) we would also require the user to finish onboarding. So a user who has signed up but hasn't onboarded could do everything on the site, but as soon as they try to download something, they are taken back into the onboarding workflow because that hasn't been finished.

carlmw commented:
> ### Testing Results
> I've tested this on the preview branch in the initial comment of this PR. Three possible issues seen:
> 
> 1. Visible flash (mentioned in this PR).
This is something we'll be improving in some upcoming work. At the moment some platform limitations need to be overcome before we can get rid of the jankyness.
> 2. Onboarding loop (seen in other testing).
Yeah this is intentional, they can navigate away after completing the onboarding process. We can adjust this behaviour if it's not desirable from a product perspective. 
> 3. Missing icon on error state.
Yes this is unavoidable unless you'd previously loaded the icon it won't be loaded.

carlmw commented:
> > > > ### Testing Results
> > > > I've tested this on the preview branch in the initial comment of this PR. Three possible issues seen:
> > > > 
> > > > 1. Visible flash (mentioned in this PR).
> > > > 2. Onboarding loop (seen in other testing).
> > > > 3. Missing icon on error state.
> > > > 
> > > > ## Visible flash (mentioned in this PR)
> > > > As mentioned in the initial comment _"There is a visible jump when the user is redirected to the onboarding screen immediately after signing in."_, this is still happening at this point of development.
> > > > ## Onboarding loop (seen in other testing)
> > > > If the user logins in and gets to the /onboarding page, but then subsequently clicks away, they are returned to /onboarding moments later and don't seem to have any way to navigate away.
> > > > ## Missing icon on error state.
> > > > I think this one was only due to the method of testing, having no network connection to simulate a disconnect. As such, the icon couldn't load but on checking the SVG was present on the relevant URL.
> > > > ![SCR-20240812-kyrj](https://private-user-images.githubusercontent.com/167184734/357086523-c8f9d173-11c6-4ae2-9dcf-9736b1e9e5af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NzM0MDksIm5iZiI6MTcyMzQ3MzEwOSwicGF0aCI6Ii8xNjcxODQ3MzQvMzU3MDg2NTIzLWM4ZjlkMTczLTExYzYtNGFlMi05ZGNmLTk3MzZiMWU5ZTVhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxMlQxNDMxNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MjdjMmU3MzRkYTFjZGJmZTE2MjczMThiODJiYTA2MTUyMTliZjVjNjgzM2M3YTVkZmM2NzU4YjFlNGZhZjE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-fwOQJxKlvyg0GsNK1YSlZuWLP3cN3k2DtCVSEkSg6c)
> > > 
> > > 
> > > @carlmw that onboarding loop is particularly problematic because it locks the user from ever using the site (without doing some things that I doubt our typical user would want to do) if they decide they don't want to continue the sign up journey. I am assuming it is part of creating the auth token so it needs to be done before they can carry on their journey?
> > 
> > 
> > @jack-skerrett-oak @mikeritson-oak at that point they are already signed-up and logged in. The intention there is to ensure that they complete the onboarding process before continuing to use the rest of the site. It's what i'm alluding to [here in Slack](https://oaknationalacademy.slack.com/archives/C03HZ76AE3Z/p1723197888223549). If you think that isn't correct we should probably raise it over on that thread as it's a product decision (I know there is a desire to capture this info). We could tweak it to only push the user through onboarding again when they try to access another page that requires authentication if that makes sense?
> 
> @carlmw that is how I interpreted that thread. I don't think we want to block users from accessing areas of the site that don't require signup because they haven't finished onboarding, but everywhere that does require signup to access (currently downloads) we would also require the user to finish onboarding. So a user who has signed up but hasn't onboarded could do everything on the site, but as soon as they try to download something, they are taken back into the onboarding workflow because that hasn't been finished.

@jack-skerrett-oak gotcha, that isn't the behaviour I was trying to allude to 🤣 . Perhaps i'm planning a little too far ahead when sign-in is more of a global concern. For now i'm probably just as comfortable limiting onboarding to the pages that we actually require it on and we can worry about the "signed-in on a page that doesn't need auth" case later on.

WIll run it past Jo when she is in.

Probably worth noting for future scenarios that there will be a "non-onboarded" user type to consider.

jack-skerrett-oak reviewed (changes requested):
Whilst we wait for Jo's verdict, there is something else.

Currently the network/server error message isn't accessible. I think we should add `aria-live=polite` to the error message itself to let screen reader users know that the error has occurred (currently you click continue and the user isn't told anything). Then we could also add `aria-describedby` onto the continue button, referencing the error so users understand the relationship between the dimmed continue button and the error that has appeared.

carlmw commented:
@jack-skerrett-oak @mikeritson-oak thanks for the thorough testing notes 🙇🏻 

1. I've updated the error element with `aria-live=polite` and given the Continue button an `aria-description` matching the error message when in the error state. I gave it a quick test with voice over and this appears to work as expected.
2. The visible flash is now reduced thanks to the changes in the next point 👇🏻 
3. I've adjusted the onboarding redirect behaviour so that signed-in users will now only be required to onboard when trying to access the authenticated download page. So users can abandon the download after signing-up and will no longer be expected to onboard until they try to download again.

jack-skerrett-oak reviewed (approved):
No flicker, I can still browse the platform when I abandon onboarding, erroring is now accessible. Looks great :)

carlmw commented on .env.test:
```diff
@@ -11,6 +11,12 @@ NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME="NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME"
 NEXT_PUBLIC_CLOUDINARY_SECURE_DISTRIBUTION="NEXT_PUBLIC_CLOUDINARY_SECURE_DISTRIBUTION"
 NEXT_PUBLIC_OAK_ASSETS_HOST="NEXT_PUBLIC_OAK_ASSETS_HOST"
 NEXT_PUBLIC_OAK_ASSETS_PATH="NEXT_PUBLIC_OAK_ASSETS_PATH"
+CURRICULUM_API_URL="http://api.example.com"
+CURRICULUM_API_2023_URL="http://api.example.com"
+CURRICULUM_API_AUTH_TYPE="auth_type"
+CURRICULUM_API_AUTH_KEY="very secret"
+CURRICULUM_API_2023_AUTH_KEY="very secret"
+
```
These being missing was causing my tests for the new endpoint to throw

carlmw commented on src/middleware.ts:
```diff
@@ -0,0 +1,11 @@
+
+export const config = {
+  matcher: [
+    // Skip Next.js internals and all static files, unless found in search params
+    // "/((?!_next|[^?]*\\.(?:html?|css|js(?!on)|jpe?g|webp|png|gif|svg|ttf|woff2?|ico|csv|docx?|xlsx?|zip|webmanifest)).*)",
```
Bypassing the clerk middleware on all routes except the one I actually need it on right now. Running the middleware on all routes was causing the middleware to rewrite the `redirect_url` param to something under the `netlify.app` deploy preview domain. Apparently the AI team faced a similar issue, so i'll check with them when I can to get a proper fix so that the middleware can be run everywhere. For now we can get by without using it.

Working this out cost me so much time today 😭

carlmw commented on src/components/TeacherComponents/OnboardingForm/OnboardingForm.tsx:
```diff
@@ -51,17 +54,27 @@ const OnboardingForm = ({
   const utmParams = useUtmParams();
   const { posthogDistinctId } = useAnalytics();
   const { user } = useUser();
+  const [submitError, setSubmitError] = useState<string | null>(null);
```
Storing this as local state so that the form can be re-submitted should the onboarding request fail.

carlmw commented on src/components/TeacherComponents/OnboardingForm/OnboardingForm.tsx:
```diff
@@ -91,6 +104,15 @@ const OnboardingForm = ({
+      // Return the user to the page they originally arrived from
+      // or to the home page as a fallback
+      router.push(
+        toSafeRedirect(
```
Sanitises the `returnTo` param from phishing attempts. Clerk should be doing this internally for `redirect_url` but since we're doing this redirect ourselves it's out responsibility.

carlmw commented on src/app/api/auth/onboarding/route.ts:
```diff
@@ -0,0 +1,17 @@
+    return new Response("Unauthorized", { status: 401 });
+  }
+
+  const publicMetadata = { "owa:onboarded": true };
```
We can add add `region` for geo location and mark the user as a teacher here.

carlmw commented on src/components/TeacherComponents/OnboardingForm/OnboardingForm.tsx:
```diff
@@ -51,17 +54,27 @@ const OnboardingForm = ({
             : "onboarding-role-selection",
         }),
-      );
+        query: router.query,
```
Maintains the `returnTo` param so that we know which page to return the user to after they have onboarded

carlmw commented:
@jack-skerrett-oak @njblore back again with Auth0 swapped for Clerk.

jack-skerrett-oak reviewed (approved):
Signing up requires onboarding, which can be abandoned and revisited later. Once signed in I don't need to do anything to access the downloads page. If signed out (but signed up), after signing in I see the downloads page without being required to onboard.

njblore commented:
When I sign up after going directly to the downloads page by URL (vs clicking through to downloads from the lesson page) I am redirected to the homepage instead of downloads page - is that expected behaviour?

carlmw commented:
> When I sign up after going directly to the downloads page by URL (vs clicking through to downloads from the lesson page) I am redirected to the homepage instead of downloads page - is that expected behaviour?

@njblore no it's not, that was the broken behaviour the middleware was causing. I thought i'd fixed it. I've just tried to replicate what you're seeing and I can't. It worked as expected after both navigating through the lesson page and with a direct link to the download page. Could you do a video or share the steps you took to get into that state?

carlmw commented:
> @carlmw [here's a loom](https://www.loom.com/share/434662a845664b21ac0bc558c1f6b506?sid=f2d3c6d5-647a-4c0a-986a-474ed537caf9)! The downloads URL I navigated to after signing out was https://deploy-preview-2681--oak-web-application.netlify.thenational.academy/teachers/programmes/art-secondary-ks3-l/units/pop-art-modern-popular-culture-4928/lessons/chila-kumari-burman-part-1-cdh3jc/downloads-auth?preselected=all Is it to do with the `-auth` in the downloads URL I navigated to?

@njblore  thanks! Now i'm properly baffled. I managed to replicate what you saw exactly once by following that link  ☝🏻 but when I tried to do it again it works fine 😮‍💨

njblore commented:
@carlmw hmm, very strange, I can have a look locally as it's happening consistently on my machine - I reckon we write a separate bug ticket though and move this along if everything else is good?

carlmw commented:
> @carlmw hmm, very strange, I can have a look locally as it's happening consistently on my machine - I reckon we write a separate bug ticket though and move this along if everything else is good?

Sounds good

njblore reviewed (approved):
🤌

oak-machine-user commented:
:tada: This PR is included in version 1.524.0 :tada:

The release is available on [GitHub release](https://github.com/oaknational/Oak-Web-Application/releases/tag/v1.524.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
